File: SISO_OFDM.m
clear all; 
close all; 
clc;
SNR_dB = [1:5:60];  %SNR range in dB
SNR = 10.^(SNR_dB/10);  %SNR in linear scale
BER = zeros(length(SNR),1);
 
Nsub = 64;  %Number of subcarriers
Ncp = round(Nsub/10);   %Number of samples added as CP
 
numTaps = 3;    %Number of channel taps
numBlocks = 10000;
 
for L = 1:numBlocks 
    Bits = randi([0,1],1,Nsub);    %Generating message bits
    h = 1/sqrt(2) * (randn(1,numTaps) + 1j*randn(1,numTaps));   
    Hfreq = fft(h, Nsub);    %N point FFT of Channel taps
    ChNoise=randn(1, Nsub+Ncp+numTaps-1) + 1j*randn(1, Nsub+Ncp+numTaps-1);

    for K = 1:length(SNR)
        LoadedBits = sqrt(SNR(K))*(2*Bits - 1);    %BPSK symbols
        TxSamples = ifft(LoadedBits);       %IFFT of symbols at transmitter
        TxSamples_CP = [TxSamples(Nsub-Ncp+1:Nsub) TxSamples];   %Adding CP 
        Rxbits = conv(TxSamples_CP, h) + ChNoise;    
        RxbitsWithoutCP = Rxbits(Ncp+1:Ncp+Nsub);   %Removal of CP
        RxbitsFFT = fft(RxbitsWithoutCP);     %FFT of received symbols at receiver
        ProcessedBits = RxbitsFFT./Hfreq;     %Equalization at receiver
        DecodedBits = (real(ProcessedBits)>=0);
        BER(K) = BER(K) + sum(DecodedBits ~= Bits);   %BER for Kth SNR value
        
    end
end
 
BER = BER/(numBlocks*Nsub);
semilogy(SNR_dB, BER, 'b - s', 'linewidth', 2.0);
hold on; 
axis tight; 
grid on;
 
SNR_eff = numTaps/Nsub * SNR;
semilogy(SNR_dB, (1/2*(1-sqrt(SNR_eff./(SNR_eff+2)))), 'r  s', 'linewidth', 2.0);
xlabel('SNRdB')
ylabel('BER')
title('BER vs SNR for SISO OFDM System');
legend('Simulation', 'Theory')

File: SISO_OFDM.py
import numpy as np
import numpy.random as nram 
import matplotlib.pyplot as plt

# Number of given subcarriers 
nSubcarriers=64
nblocks=10000

# Variance
No = 1

Eb_db=np.arange(1,61,6)
Eb= 10**(Eb_db/10)
SNR=2*(Eb/No)
SNR_db=10*np.log10(SNR)
BER=np.zeros(len(Eb_db))
BER_theory=np.zeros(len(Eb_db))

# Number of unit gain channel taps
L=3

# Number of symbols used for cyclic prefix 
L_tilde=6

for i in range(nblocks):
    # Generation of inphase bits and quadrature bits
    inphase_bits = nram.randint(2,size=nSubcarriers)
    quad_bits = nram.randint(2,size=nSubcarriers)

    # Generation of QPSK symbols
    Qpsk_sym = (2*inphase_bits-1)+1j*(2*quad_bits-1)

    # Generation of noise for both inphase and quadrature bits
    inphase_noise = nram.normal(0,np.sqrt(No/2),nSubcarriers+L-1+L_tilde)
    quad_noise = nram.normal(0,np.sqrt(No/2),nSubcarriers+L-1+L_tilde)

    

    # Complex noise
    noise = inphase_noise+1j*quad_noise

    # Unit gain channel taps
    h=nram.normal(0,np.sqrt(1/2),L)+1j*nram.normal(0,np.sqrt(1/2),L)

    # FFT of unit gain channel taps
    fft_of_h = np.fft.fft(h,nSubcarriers)

    for j in range(len(Eb_db)):
        # Tramitted symbols
        x_of_n=np.sqrt(Eb[j])*Qpsk_sym

        # IFFT of transmitted symbols
        tx_sym=np.fft.ifft(x_of_n)

# Copying the last L_tilde symbols from tansmitted symbols for cyclic   prefix
        cyclic_pref=tx_sym[nSubcarriers-L_tilde:]

        # Addition of cyclic prefix
        tx_sym_cyclic_pref=np.concatenate((cyclic_pref,tx_sym))

        # Received symbols
        rx_sym_cyclic_pref=np.convolve(tx_sym_cyclic_pref,h)+noise

        # Removal of cyclic prefix
        rx_sym=rx_sym_cyclic_pref[L_tilde:L_tilde+nSubcarriers]

        # FFT of received symbols 
        fft_of_rx_sym=np.fft.fft(rx_sym)

        # Single tap equalisation
        sin_tap_equalisation=1/fft_of_h*fft_of_rx_sym

        # Detection of inphase and quadrature bits
        det_inphase_bits=(np.real(sin_tap_equalisation)>0)
        det_quad_bits=(np.imag(sin_tap_equalisation)>0)

        # BER Simulation
        BER[j]=BER[j]+np.sum(det_inphase_bits!=inphase_bits)+np.sum(det_quad_bits!=quad_bits)

BER=BER/nSubcarriers/2/nblocks

# Theoretical BER calculation
# Effective SNR
effective_SNR = (L*SNR)/nSubcarriers



# BER Theoretical
BER_theory = 0.5*(1-np.sqrt(effective_SNR/(2+effective_SNR)))


# Superimposed plot BER theoretical and BER simulation
plt.figure(1)
plt.yscale('Log')
plt.plot(SNR_db,BER,'r')
plt.plot(SNR_db,BER_theory,'k:o')
plt.grid(1,which='both')
plt.suptitle('BER for OFDM wireless system')
plt.xlabel('SNR(in dbs)')
plt.ylabel('BER')
plt.legend(["BER Simulation","BER THEORY"])
plt.show()



